framework/transport/ieee1905_transport/ieee1905_transport_packet_processing.cpp

```diff
@@ -577,6 +579,64 @@ bool Ieee1905Transport::forward_packet(Packet &packet)
     return true;
 }
 
+/**
+ * @brief forward a 1905.1 packet.
+ * 
+ * Basically a wrapper to forward_packet_single, which also takes into account the reliable
+ * multicast procedure - according to section 15.1 in the Multi-AP specification, a reliable multicast
+ * transmission includes sending the packet with the relay bit set to a multicast address once,
+ * and once for each discovered Multi-AP agent in the network as unicast
+ * with the same MID and the relay bit unset.
+ * 
+ * @param packet packet to forward
+ * @return true on success
+ * @return false on failure (one or more packets failed in transmission)
+ */
+bool Ieee1905Transport::forward_packet(Packet &packet)
+{
+    // First, forward the packet as is.
+    bool success = forward_packet_single(packet);
+    if (!success) {
+        MAPF_DBG("Failed to forward packet " << packet);
+        return false;
+    }
+
+    // Check if this packet should also go through reliable multicast:
+    // 1. Destination is multicast
+    // 2. Message type is topology notification, which is currently the only
+    //    message which should be sent using reliable multicast. See Multi-AP
+    //    specification Table 5
+    auto should_send_reliable_multicast = [&](const Packet &packet) -> bool {
+        Ieee1905CmduHeader *ch = (Ieee1905CmduHeader *)packet.payload.iov_base;
+        return ETHER_IS_MULTICAST(packet.dst.oct) && ch->GetRelayIndicator() &&
+               ntohs(ch->messageType) ==
+                   static_cast<uint16_t>(ieee1905_1::eMessageType::TOPOLOGY_NOTIFICATION_MESSAGE);
+    };
+    if (!should_send_reliable_multicast(packet)) {
+        return true;
+    }
```

Comment:
Same comment as @kantera800 made, but please explain why this is more readable.
```suggestion
    // Check if this packet should also go through reliable multicast, return if not:
    // 1. Destination is multicast
    if (!ETHER_IS_MULTICAST(packet.dst.oct)) {
        return true;
    }
    // 2. Message is relayed multicast
    Ieee1905CmduHeader *ch = (Ieee1905CmduHeader *)packet.payload.iov_base;
    if (!ch->GetRelayIndicator()) {
        return true;
    }
    // 3. Message type is topology notification, which is currently the only
    //    message which should be sent using reliable multicast. See Multi-AP
    //    specification Table 5
    if (ntohs(ch->messageType) != static_cast<uint16_t>(ieee1905_1::eMessageType::TOPOLOGY_NOTIFICATION_MESSAGE)) {
        return true;
    }
```

Also, you forgot one condition: we should only do this if the message was generated locally, not if we're relaying it ourselves. The spec says to duplicate the message as part of the transmission procedure (section 7.3 of 1905.1) while relaying is part of the reception procedure (section 7.6 of 1905.1).

Reply:
Its a matter of taste I guess, but I actually like your suggestion so I'll change it.